File: top_100_liked_algos_and_data_structures/last_first_occurrence/last_first.py
a = [3, 5, 6, 2, 4, 6, 7, 2, 5,7, 3, 2,7]
a.sort()

def first_occurrence(arr, target):
    left, right = 0, len(arr) - 1
    result = -1

    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            result = mid
            right = mid - 1

        elif arr[mid] < target: right = mid - 1
        else: left = mid + 1

    return result

def last_occurrence(arr, target):
    left, right = 0, len(arr) - 1
    result = -1

    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            result = mid
            left = mid + 1

        elif arr[mid] == target: right = mid - 1
        else: left = mid + 1

File: top_100_liked_algos_and_data_structures/last_first_occurrence/last_first.rs
fn first_occurrence(list: &[i32], target: i32) -> i32  {
    let (mut left, mut right) = (0, list.len() - 1);
    let mut result: i32 = -1;

    while left <= right {
        let mid = (left + right) / 2;
        if *&list[*&mid] == *&target {
            result = *&mid as i32;
            right = *&mid - 1 as usize;
        } else if *&list[*&mid] < *&target {
            left  = *&mid + 1 as usize;
        } else {
            right = *&mid - 1 as usize;
        }
    }

    return result;
}

fn last_occurrence(list: &[i32], target: i32) -> i32 {
    let (mut left, mut right) = (0, list.len() - 1);
    let mut result: i32 = -1;

    while left <= right {
        let mid = (left + right) / 2;

        if *&list[*&mid] == *&target {
            result = *&mid as i32;
            left = *&mid + 1 as usize;
        }
        else if *&list[*&mid] < *&target {
            right = *&mid - 1 as usize;
        }
        else {
            left = *&mid + 1 as usize;
        }
    }

    return result;
}

fn main() {
    let arr: [i32;13] = [2, 2, 2, 3, 3, 4, 5, 5, 6, 6, 7, 7, 7];
    println!("{}", first_occurrence(&arr, 5));
}